### ISSUE TITLE
Funky looking cells in the fits image

### ISSUE THREAD

**Issue by cleevesli** [lime-rt/lime#162]
Hey all,

I cloned yesterday's version of LIME and have been putting in my own model reading code bits back in, and ran identical models before and after the update to compare. I've noticed in the most recent versions (both the v1.6 commit yesterday and a few days ago) that there are now giant cells in the high velocity line channels for both the red and blue shifted emission (see attached images). Any thoughts on what could be causing this? The vtk grid files as viewed with ParaView look fine, and the fluxes for the circumstellar disk emission (the center bowtie part) are very similar to the previous version. 

Best,
Ilse

<img width="477" alt="screen shot 2016-09-03 at 6 51 42 am" src="https://cloud.githubusercontent.com/assets/6672240/18224658/f5ab1570-71a2-11e6-8e44-0a55aa7e9858.png">

<img width="458" alt="screen shot 2016-09-03 at 6 52 06 am" src="https://cloud.githubusercontent.com/assets/6672240/18224659/f9892056-71a2-11e6-93f7-f19e905153de.png">

**Comment by cleevesli** [comment]
I also just noticed if I do an extremely narrow stretch on the color bar range that the background is a little strange looking too, with the black parts being 1e-30 and the grey being exactly 3.21e-5:

<img width="710" alt="screen shot 2016-09-03 at 6 57 18 am" src="https://cloud.githubusercontent.com/assets/6672240/18224682/b3495fec-71a3-11e6-971a-b936874f4291.png">

Best,
ilse

**Comment by smaret** [comment]
Hey Ilse !
Do you expect any gas emission in these high velocity channels or dust emission only?  The way the dust density computation has changed in v1.6 (see the discussion in #128), so you may what to check this.

**Comment by cleevesli** [comment]
Hello Sebastian!

Last night I started backtracking through commits to see where the issue first appears and was able to trace it back to the large August 1st commit: 

https://github.com/lime-rt/lime/commit/e9956e8d409e21a26144b5e306bdd9a519c5af2b

It is not present in the July 20th commit:

https://github.com/lime-rt/lime/commit/9eaa28365db1403277606d20aaf7c97bdd053092

And I haven't yet gotten the July 21st commit running with my chemical model inputs, but August 1st is definitely not working. 

Regarding line vs. dust, it's definitely in the line emission, and to check I re-ran the model without dust and the artifacts in the fits image are still there.

I don't know exactly where the issue appears within the August 1 version, but I'll start backtracking through that one next..

Best,
Ilse

**Comment by tlunttil** [comment]
You can see the same problem with the example model if you increase the image size to e.g. 500 pixels.

I think it's an artefact of image plane interpolation. Not sure how to best fix it, though...  Any thoughts @allegroLeiden ?

**Comment by cleevesli** [comment]
Just as a quick update, the strange looking emission appears in the commit on July 22nd: https://github.com/lime-rt/lime/commit/6e81279e5f119beb1141d2494851d330ec27ebee
Glad I hadn't started working through August 1 yet. :)

**Comment by cleevesli** [comment]
@tlunttil  I was only using 250 pixels for the images above, though admittedly quite a few channels:

    img[0].nchan      = 650;         // Number of channels 
    img[0].velres     = 33.3586;      // Channel resolution in m/s 
    img[0].trans      = 1;      // zero-indexed J quantum number 
    img[0].pxls       = 250;        // Pixels per dimension 
    img[0].imgres     = 0.03;       // Resolution in arc seconds 
    img[0].theta      = 1.308;      // 0: face-on, pi/2: edge-on 
    img[0].distance   = 145.0*PC;   // source distance in pc  
    img[0].source_vel = 0;          // source velocity in m/s 
    img[0].unit       = 1;          // 0:Kelvin 1:Jansky/pixel 2:SI 3:Lsun/pixel 4:tau

**Comment by imckstewart** [comment]
Yeah this looks like my code - it would be the one non-BWC change we made. :(

There should be no change to the basic solution, what has changed is (as @tlunttil said) is the final image-plane interpolation that we reinstated to speed up the raytracing, as well as making the images smoother. We (or more accurately, I) decided not to keep the old alternative on grounds of reducing the maze of choices a little. That may have been a mistake... I'll look at this when I can but a bunch of sysadminy type things which I have been putting off descended on me post-1.6 in a rush yesterday and it may be a while before I shake free... and then starting next week I'll be vacationing in Turkey for a couple of weeks.

**Comment by imckstewart** [comment]
> The way the dust density computation has changed in v1.6

There has of course been extensive rearrangement of the appropriate code to increase both the robustness as well as the user choices, but I tried to keep the default behaviour of the dust density calculation etc exactly the same as in 1.5, and I believe I succeeded. I'd be surprised if there was a difference here - can't 100% rule it out, but I'd be surprised.

**Comment by tlunttil** [comment]
The problem looks to be that in some cases the current point where the code tries to calculate the interpolated value ends up outside the 'best' triangle found by qhull. Then interpolation turns into extrapolation and strange things can happen. 

One could think that this would be easily fixed by clamping to zero outside the triangle (which can be done easily by checking if any of the barycentric coordinates is <0). Unfortunately this doesn't work: the final image will be full of ugly black stripes (for some strange reason that has something to do with qhull...).

**Comment by imckstewart** [comment]
Bloody qhull...

**Comment by cleevesli** [comment]
As a science user who would like to start a grid of LIME models, is the last normally working commit (July 20) https://github.com/lime-rt/lime/commit/9eaa28365db1403277606d20aaf7c97bdd053092 robust for analysis? Or should I wait for the raytracing speed-ups?  Smooth images are nice but not required...

**Comment by imckstewart** [comment]
> is the last normally working commit (July 20) 9eaa283 robust for analysis?

Er - probably. We certainly tried to keep things working with each commit. Otherwise you could use 1.5, if you don't mind the flipped X axis or the (more problematic) flip in the velocity direction.

**Comment by imckstewart** [comment]
Could you send me your model file? With luck I will get a bit of time to look into this tomorrow.

**Comment by cleevesli** [comment]
Hey Ian, sure, though according to @tlunttil it's also present in the default model.c test. The model file (run2.c) along with the required input .dat file and weights.c file are in the attached zip. 
[modfiles.zip](https://github.com/lime-rt/lime/files/456942/modfiles.zip)

You'll also need to add the following to lime.h: 

```
/*Added this to include RZ structure as global*/
#define modpoints 40000   /* maximum number of points in model */
double ctrchm;
double arrchem[modpoints][6];
double arrchemfg[modpoints][7];
double maxdenval;
double mindenval;
double maxabunval;
double minabunval;
double isoabund;
int modelfilepoints;
double MINRAD,enhab,arrsz;
double stellarmass;
int wtcount;
int chemread();
```

And then to main.c right after the call to parseInput():
`chemread();`

I turned off the dust and set gasIIdust to a constant to make it easier and quicker to run.

Let me know if anything breaks!  Thanks!

- Ilse

**Comment by imckstewart** [comment]
Thanks. I've been discussing the cause and possible fixes offline with Tuomas.

**Comment by imckstewart** [comment]
Ilse, your code wouldn't run as-is. I had to make the following mods:
  - grid.c:695 - changed to match your pointEvaluation() interface (also the function prototype in lime.h).
  - run2.c:425 - semicolon at end of line.

After running it, I could not duplicate your example images above. Without a fixed random number seed, exact duplication would not be expected, but I could not come very close either. Which channels correspond to your images?

However, I do observe features (not all are bugs :) as follows:
  - Giant distal cells.
      This is normal. In particular, setting `par->sampling=2` results in the density of grid cells being very concentrated towards the centre. There is still IMO no good grid-point generation mechanism in Lime - there is one in the works, but it won't appear til 1.7.

  - 1e-30 in outer regions.
      I have always seen this in the output of Lime, and had assumed it was as originally hard-wired into the code, since 1e-30 is used as a default for other values, for reasons best known to the original developer; but now I come to look into it, a cursory examination suggests the background value ought to be set to zero. I'll need to dig a bit deeper for this one. In any case, Lime at present is not handling the outside-model rays in a graceful manner, irrespective of problems with qhull (see below). We need to work on this.

  - Constant value (I get 8.02e-6 rather than 3.21e-5) in distal cells.
      Without digging I am not sure - CMB? Low opacity in the outer regions?

  - Circular edges to distal cells.
      This is a bug, with cause resting in the package qhull as identified by @tlunttil, see his comments above. I have a fix which ought to work, but I am not sure if I can code it up before I leave for my vacation.

**Comment by cleevesli** [comment]
Thanks for looking into this! Those features seem broadly similar to what I was seeing, the background value, circular outer cell edges, non-zero background... Some changes are probably due to the fact that the images I attached originally were from the most recent 1.6 release of LIME, and the scripts I sent were for the last "working" version of LIME for me, July 20th. The most concerning thing for me are the large extended (spiky) cells that stretch out far beyond the expected emitting region (are these the distal cells you're referring to?). Those are visible in all of my models for LIME versions post July 20th. They appear in the channels that have the most radially extended emission in the projected image plane, not at the line center or in the very high velocity wings.

**Comment by tlunttil** [comment]
@allegroLeiden : that 1e-30 comes from line 123 of writefits.c. As for why the values are clamped to 1e-30, I have no idea.

**Comment by cleevesli** [comment]
If you'd like to see the full output, here is an unnecessarily giant (400 mb) fits cube:

https://www.dropbox.com/s/uefh04782ogk3l3/bhb_gap2oom_v2_CO_1_0p3177_230.fits?dl=0

**Comment by imckstewart** [comment]
The spiky cells are anomalous, there is no doubt about it, but it is their shape which is the problem rather than their size. You will get big cells in the outfield unless you (i) modify the grid-cell sampling function so it is less peaked or (ii) specify more grid points.

**Comment by cleevesli** [comment]
So to confirm -- when you run my model, do you also see the spiky cells as having anomalous flux? (I understand that they are intrinsically part of the sampling).  I was using the old qhull flag as well.

**Comment by imckstewart** [comment]
Sorry, I guess I am being dense but this is not enough information. What do you mean by 'anomalous flux'? I am also puzzled by 'old qhull flag' because I was not aware that we had changed it.

I am currently testing a fix but if I can't figure out why it is seg faulting this will probably not get patched before I come back from holiday, sorry.

**Comment by tlunttil** [comment]
I think old qhull means the thing in the Makefile.

**Comment by cleevesli** [comment]
Yeah, qhull versus qhullstatic.

I meant do you see similar features to the original images I posted, i.e.,
the giant bright spikes coming off the disk?  The disk itself (the central
part) looks completely normal compared to old versions, just the spikes are
new as of the July 20th changes.

L. Ilsedore Cleeves
Hubble Fellow, Harvard-Smithsonian Center for Astrophysics
60 Garden St, MS-51
Cambridge, MA 02138
1-706-237-8398
ilse.cleeves@cfa.harvard.edu
https://www.cfa.harvard.edu/~lcleeves/index.html

**Comment by tlunttil** [comment]
Should the model take very long time to run?

**Comment by cleevesli** [comment]
it should take 25 minutes before the gasIIdust bug was fixed and ~ 3 minutes after.

**Comment by tlunttil** [comment]
Yes, I can reproduce the problem and it seems to be caused by some issues in interpolation, like I guessed yesterday.

**Comment by allegroLeiden** [comment]
Ilse do you want to try

  https://github.com/allegroLeiden/lime/archive/fix_162_a.zip

**Comment by allegroLeiden** [comment]
'qhull flag' - ok I was thinking of the Qbb etc stuff in the call to qhull. D'oh!!

**Comment by cleevesli** [comment]
@allegroLeiden works like a charm, and the fluxes are in good agreement with my previously working version. thank you! this is v1.6 yes? 

The background also looks more normal now (see attached stretched image), back to a circle with the radial size of the source with a constant value of 8.02e-6. Not really an issue for my purposes since it's value is so small compared to the source, but if this is supposed to be the CMB shouldn't it fill the entire image? (though this question probably should be another ticket...). 

<img width="262" alt="screen shot 2016-09-08 at 8 13 10 am" src="https://cloud.githubusercontent.com/assets/6672240/18348902/245bfb08-759c-11e6-890d-5f305bb7e30c.png">

**Comment by allegroLeiden** [comment]
Hi @cleevesli , yes I modified 1.6 for this. It is just a few extra lines in raytrace.c. I'll put in a pull request, so hopefully it will appear as a patch sometime soon - exactly when will depend on the other developers, since I'll be away for 2 weeks.

Actually the interpolation still seems not quite right, I have a plan B to try but that will also have to wait, since it involves more elaborate changes.

> if this is supposed to be the CMB shouldn't it fill the entire image?

Image pixels which don't overlap the model sphere are simply left at zero in the raytracing routine - it is later on when the cube is written to file that all low values are clipped at 1e-30. I don't know why this was done originally, and at some stage we will probably change the output indeed as you suggest, but there is a long list of structural and algorithmic issues to address in Lime before we can get to such relatively minor points. I for one will not begin to be truly happy with Lime until 1.7 and will be pushing towards that release just as fast as I can manage it.

**Comment by tlunttil** [comment]
I'm not sure if setting pixels outside the model area to cmb (or any other particular value) is necessarily a good idea. Maybe it would be better to set them to NaN?

**Comment by allegroLeiden** [comment]
TBD

**Comment by smaret** [comment]
Thanks @cleevesli for reporting this bug and to @allegroLeiden and @tlunttil for fixing it. I'm going to merge Ian's fix  (#165) and close this issue. Please open a new one of the CMB / 1e-30 values (I agree this is something we should look into).

**Comment by smaret** [comment]
FYI: I've just made a [bug fix release (1.6.1)](https://github.com/lime-rt/lime/releases/tag/v1.6.1) with Ian's fix.

**Comment by cleevesli** [comment]
Thanks everyone! So regarding @allegroLeiden's comment

> Actually the interpolation still seems not quite right, I have a plan B to try but that will also have to wait, since it involves more elaborate changes.

Is this now solved?

**Comment by tlunttil** [comment]
> Is this now solved? 

Not quite. There's still a problem with interpolation, but it is not as serious as before the latest patch. 

I'm not sure what exactly @allegroLeiden 's  plan B is and he's on vacation, so it may be be a while before he replies. I could write a quick-and-dirty fix, but I'd rather not do that before I've heard about @allegroLeiden 's plans (unless that is something you need urgently).

**Comment by cleevesli** [comment]
@tlunttil What errant behavior should I be looking for?

**Comment by tlunttil** [comment]
Basically the same kind of behaviour as in your original bug report, but on a smaller scale (smaller erratic cells).

**Comment by allegroLeiden** [comment]
I am not sure why this was closed...

My 'plan B' was to adapt the bespoke code already in use for finding the chain of Delaunay tetrahedra traversed by any ray to 2D. I.e., working in the image plane we have not tetrahedra but triangles, and instead of a ray or light path we have a raster across the image. One would, for each row of the image, start at the leftmost column, find the chain of triangles traversed by the row (thus avoiding the flaky qhull routine), then work along the row doing the interpolation as desired. I'll try to code this up and try it next week.

**Comment by smaret** [comment]
> I am not sure why this was closed...

I thought that your PR #165 (which I've merged) fixed this issue, so I closed it. Please feel free to re-open it if necessary.

**Comment by allegroLeiden** [comment]
'fixed'->'improved'.

**Comment by allegroLeiden** [comment]
I have been looking at making the code which follows rays through the mesh of Delaunay cells more generic - at the moment there is fairly copious hard-wiring of the number of spatial dimensions to 3, or at least dependence on the global macro DIM. The idea was that, if it could be made generic, with the number of dims passed in as a parameter, which is much better practice anyway, I could use the same functions in 2D to follow image raster lines through the mesh of projected ray locations. However, I have found it is not so easy, and would require some rejigging of the way model points are stored and so forth. Some of us have been thinking of doing something like this anyway, but in terms of scale of modification we are talking about much more than just a patch.

So I think the best thing would be if I postpone my 'plan B' for now and for @tlunttil (or somebody else?) to implement @tlunttil 's brute-force workaround for the patch which will hopefully close this issue.

**Comment by tlunttil** [comment]
Unfortunately I'm busy with all sorts of things right now. I can write a PR for 1.7, but if a separate patch to fix this is needed soon, it probably needs to be done by someone else.

**Comment by allegroLeiden** [comment]
This was fixed in 1.6.1.

**Comment by allegroLeiden** [comment]
This was fixed in 1.6.1.
